File: solutions/lian-biao-zhong-dao-shu-di-kge-jie-dian-lcof/lian-biao-zhong-dao-shu-di-kge-jie-dian-lcof.go
// @Title: 链表中倒数第k个节点 LCOF
// @Author: colinjxc
// @Date: 2021-07-20T12:42:11+08:00
// @URL: https://leetcode-cn.com/problems/lian-biao-zhong-dao-shu-di-kge-jie-dian-lcof


/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func getKthFromEnd(head *ListNode, k int) *ListNode {
    fast := head
    slow := head
    for k > 0 && fast != nil {
        fast = fast.Next
        k--
    }

    for fast != nil {
        fast = fast.Next
        slow = slow.Next
    }

    return slow
}

File: solutions/lian-biao-zhong-dao-shu-di-kge-jie-dian-lcof/lian-biao-zhong-dao-shu-di-kge-jie-dian-lcof.cpp
// @Title: 链表中倒数第k个节点 LCOF
// @Author: colinjxc
// @Date: 2022-03-01T09:35:36+08:00
// @URL: https://leetcode-cn.com/problems/lian-biao-zhong-dao-shu-di-kge-jie-dian-lcof


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* getKthFromEnd(ListNode* head, int k) {
        ListNode *fast = head, *slow = head;
        for (int i = 0; i < k && fast; i++) {
            fast = fast->next;
        }
        while (fast && slow) {
            fast = fast->next;
            slow = slow->next;
        }

        return slow;
    }
};
